File: backend/routes/quality/faculty.py
from flask import Blueprint, jsonify, request
from functools import wraps
from supabase_client import get_supabase

# Create blueprint
quality_faculty_bp = Blueprint('quality_faculty', __name__)

@quality_faculty_bp.route('/faculty', methods=['GET', 'OPTIONS'])
def get_faculty():
    """Get faculty list with pagination from Supabase"""
    try:
        page = int(request.args.get('page', 1))
        limit = int(request.args.get('limit', 10))
        search = request.args.get('search', '')
        department = request.args.get('department', '')

        supabase = get_supabase()

        # Build query
        query = supabase.table('quality_facultyperformance').select('*', count='exact')

        # Apply filters
        if search:
            # Search in faculty_name (case insensitive)
            query = query.ilike('faculty_name', f'%{search}%')
        if department:
            query = query.eq('department', department)

        # Get results
        offset = (page - 1) * limit
        result = query.range(offset, offset + limit - 1).execute()
        total_items = result.count if hasattr(result, 'count') else 0

        # Transform data to match frontend expectations
        faculty = []
        for record in result.data:
            faculty.append({
                'id': record.get('faculty_id'),
                'employee_id': f"EMP{record.get('faculty_id')}",
                'name': record.get('faculty_name'),
                'email': f"{record.get('faculty_name', '').lower().replace(' ', '.')}@college.edu",
                'department': record.get('department'),
                'designation': 'Professor', # Not in quality_facultyperformance
                'performance_rating': float(record.get('performance_rating', 0)),
                'research_output': record.get('research_papers', 0),
                'student_feedback_score': float(record.get('feedback_score', 0)),
                'teaching_hours': 20,
                'publications': record.get('research_papers', 0),
                'projects': 0,
                'status': 'active'
            })

        total_pages = (total_items + limit - 1) // limit if total_items > 0 else 1

        return jsonify({
            'success': True,
            'data': faculty,
            'pagination': {
                'page': page,
                'limit': limit,
                'total': total_items,
                'totalPages': total_pages
            }
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_faculty_bp.route('/faculty/analytics', methods=['GET', 'OPTIONS'])
def get_faculty_analytics():
    """Get faculty analytics"""
    try:
        supabase = get_supabase()
        result = supabase.table('quality_facultyperformance').select('*').execute()
        data = result.data or []
        
        total_faculty = len(data)
        active_faculty = total_faculty 
        
        dept_dist = {}
        desig_dist = {'Professor': total_faculty} # Mock since not in table
        perf_dist = {'excellent': 0, 'good': 0, 'average': 0, 'needs_improvement': 0}
        total_perf = 0
        
        for record in data:
            dept = record.get('department')
            if dept:
                dept_dist[dept] = dept_dist.get(dept, 0) + 1
                
            rating = float(record.get('performance_rating', 0))
            total_perf += rating
            if rating >= 90: perf_dist['excellent'] += 1
            elif rating >= 75: perf_dist['good'] += 1
            elif rating >= 60: perf_dist['average'] += 1
            else: perf_dist['needs_improvement'] += 1
            
        avg_perf = total_perf / total_faculty if total_faculty > 0 else 0
        
        analytics = {
            'total_faculty': total_faculty,
            'active_faculty': active_faculty,
            'faculty_by_department': dept_dist,
            'faculty_by_designation': desig_dist,
            'performance_distribution': perf_dist,
            'average_performance': round(avg_perf, 1),
            'performance_trends': [
                {'month': 'Jan', 'score': 85},
                {'month': 'Feb', 'score': 87},
                {'month': 'Mar', 'score': 88},
                {'month': 'Apr', 'score': 86},
                {'month': 'May', 'score': 90},
                {'month': 'Jun', 'score': 92}
            ],
            'research_output': [
                {'month': 'Jan', 'count': 8},
                {'month': 'Feb', 'count': 12},
                {'month': 'Mar', 'count': 15},
                {'month': 'Apr', 'count': 10},
                {'month': 'May', 'count': 18},
                {'month': 'Jun', 'count': 14}
            ]
        }
        
        return jsonify({
            'success': True,
            'data': analytics
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_faculty_bp.route('/faculty', methods=['POST', 'OPTIONS'])
def create_faculty():
    """Create a new faculty member and save to Supabase"""
    try:
        data = request.get_json()
        supabase = get_supabase()

        # Map frontend fields to DB columns for quality_facultyperformance
        insert_data = {
            'faculty_name': data.get('name'),
            'department': data.get('department'),
            'performance_rating': float(data.get('performance_rating', 0)),
            'research_papers': int(data.get('research_output', 0)),
            'feedback_score': float(data.get('student_feedback_score', 0))
        }

        result = supabase.table('quality_facultyperformance').insert(insert_data).execute()

        return jsonify({
            'success': True,
            'data': result.data[0] if result.data else insert_data,
            'message': 'Faculty member created successfully'
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_faculty_bp.route('/faculty/<string:faculty_id>', methods=['PUT', 'OPTIONS'])
def update_faculty(faculty_id):
    """Update a faculty member in Supabase"""
    try:
        data = request.get_json()
        supabase = get_supabase()

        update_data = {}
        if 'name' in data:
            update_data['faculty_name'] = data.get('name')
        if 'department' in data:
            update_data['department'] = data.get('department')
        if 'performance_rating' in data:
            update_data['performance_rating'] = float(data.get('performance_rating', 0))
        if 'research_output' in data:
            update_data['research_papers'] = int(data.get('research_output', 0))
        if 'student_feedback_score' in data:
            update_data['feedback_score'] = float(data.get('student_feedback_score', 0))

        if not update_data:
            return jsonify({'success': False, 'error': 'No valid fields to update'}), 400

        result = supabase.table('quality_facultyperformance').update(update_data).eq('faculty_id', faculty_id).execute()

        return jsonify({
            'success': True,
            'data': result.data[0] if result.data else {'id': faculty_id, **update_data},
            'message': 'Faculty member updated successfully'
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@quality_faculty_bp.route('/faculty/<string:faculty_id>', methods=['DELETE', 'OPTIONS'])
def delete_faculty(faculty_id):
    """Delete a faculty member from Supabase"""
    try:
        supabase = get_supabase()

        result = supabase.table('quality_facultyperformance').delete().eq('faculty_id', faculty_id).execute()

        return jsonify({
            'success': True,
            'message': 'Faculty member deleted successfully'
        })
    except Exception as e:
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

File: Updated_CMS-main/backend/controllers/quality/faculty.js
const { supabase } = require('../../supabase_client');

// SUB-MODULE 2: FACULTY PERFORMANCE DATA
exports.getAllFaculty = async (req, res) => {
  try {
    const { page = 1, limit = 10, department, search } = req.query;
    
    let query = supabase
      .from('quality_facultyperformance')
      .select('*', { count: 'exact' });

    if (department && department !== 'All Departments') {
      query = query.eq('department', department);
    }

    if (search) {
      query = query.or(`faculty_name.ilike.%${search}%,faculty_id.ilike.%${search}%`);
    }

    const from = (page - 1) * limit;
    const to = from + limit - 1;

    const { data, error, count } = await query.range(from, to);

    if (error) throw error;

    res.json({
      success: true,
      data: data || [],
      pagination: {
        page: parseInt(page),
        limit: parseInt(limit),
        total: count || 0,
        totalPages: Math.ceil((count || 0) / limit)
      }
    });
  } catch (error) {
    console.error('Error fetching faculty:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch faculty data'
    });
  }
};

exports.addFaculty = async (req, res) => {
  try {
    // Generate faculty_id if not provided
    if (!req.body.faculty_id) {
      const { data: lastFaculty } = await supabase
        .from('quality_facultyperformance')
        .select('faculty_id')
        .order('faculty_id', { ascending: false })
        .limit(1);
      
      let nextId = 'F0001';
      if (lastFaculty && lastFaculty.length > 0) {
        const lastId = lastFaculty[0].faculty_id;
        const currentNum = parseInt(lastId.substring(1));
        nextId = `F${(currentNum + 1).toString().padStart(4, '0')}`;
      }
      req.body.faculty_id = nextId;
    }

    const { data, error } = await supabase
      .from('quality_facultyperformance')
      .insert([req.body])
      .select();

    if (error) throw error;

    res.status(201).json({
      success: true,
      data: data[0],
      message: 'Faculty added successfully'
    });
  } catch (error) {
    console.error('Error adding faculty:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to add faculty'
    });
  }
};

exports.updateFaculty = async (req, res) => {
  try {
    const { id } = req.params;
    
    const { data, error } = await supabase
      .from('quality_facultyperformance')
      .update(req.body)
      .eq('faculty_id', id)
      .select();

    if (error) throw error;

    if (!data || data.length === 0) {
      return res.status(404).json({
        success: false,
        error: 'Faculty not found'
      });
    }

    res.json({
      success: true,
      data: data[0],
      message: 'Faculty updated successfully'
    });
  } catch (error) {
    console.error('Error updating faculty:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to update faculty'
    });
  }
};

exports.deleteFaculty = async (req, res) => {
  try {
    const { id } = req.params;
    
    const { error } = await supabase
      .from('quality_facultyperformance')
      .delete()
      .eq('faculty_id', id);

    if (error) throw error;

    res.json({
      success: true,
      message: 'Faculty deleted successfully'
    });
  } catch (error) {
    console.error('Error deleting faculty:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to delete faculty'
    });
  }
};

exports.getFacultyAnalytics = async (req, res) => {
  try {
    const { data: facultyData, error } = await supabase
      .from('quality_facultyperformance')
      .select('*');

    if (error) throw error;

    // Performance trends (mock data for demo)
    const performanceTrends = [
      { month: 'Jan', score: 75 },
      { month: 'Feb', score: 78 },
      { month: 'Mar', score: 82 },
      { month: 'Apr', score: 80 },
      { month: 'May', score: 85 },
      { month: 'Jun', score: 88 }
    ];

    // Research output trends (mock data for demo)
    const researchOutput = [
      { month: 'Jan', count: 15 },
      { month: 'Feb', count: 18 },
      { month: 'Mar', count: 22 },
      { month: 'Apr', count: 20 },
      { month: 'May', count: 25 },
      { month: 'Jun', count: 28 }
    ];

    // Department comparison based on actual data
    const departmentComparison = facultyData ? 
      facultyData.reduce((acc, faculty) => {
        const deptName = faculty.department;
        const dept = acc.find(item => item.department === deptName);
        if (dept) {
          dept.count += 1;
          dept.total_score += faculty.performance_rating;
          dept.score = Math.floor(dept.total_score / dept.count);
        } else {
          acc.push({
            department: deptName,
            count: 1,
            total_score: faculty.performance_rating,
            score: faculty.performance_rating
          });
        }
        return acc;
      }, []) : [];

    // Add some stats
    const totalFaculty = facultyData ? facultyData.length : 0;
    const avgPerformance = totalFaculty > 0 
      ? Math.floor(facultyData.reduce((sum, f) => sum + (f.performance_rating || 0), 0) / totalFaculty)
      : 0;
    
    const totalResearch = facultyData ? facultyData.reduce((sum, f) => sum + (f.research_papers || 0), 0) : 0;

    const analytics = {
      performance_trends: performanceTrends,
      research_output: researchOutput,
      department_comparison: departmentComparison,
      stats: {
        total_faculty: totalFaculty,
        avg_performance: avgPerformance,
        total_research: totalResearch,
        total_publications: totalResearch // Using research_papers as publications for now
      }
    };

    res.json({
      success: true,
      data: analytics
    });
  } catch (error) {
    console.error('Error fetching faculty analytics:', error);
    res.status(500).json({
      success: false,
      error: 'Failed to fetch faculty analytics'
    });
  }
};
